From cae4f7b4c9b5d17e60755c2b2f4405183403f368 Mon Sep 17 00:00:00 2001
From: FZ_XiaoXi <FZXiaoxi@outlook.com>
Date: Tue, 11 Aug 2026 19:52:10 +0800
Subject: [PATCH 03/19] set ysyx_id

---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index cb63fd5b..94bf9765 100644
--- a/Makefile
+++ b/Makefile
@@ -24,6 +24,7 @@ verilog: verilog_fix
 	sed -i 's/              sdram_bundle_dqm/output [3:0]  sdram_bundle_dqm/' build/ysyxSoCFull.v
 	sed -i 's/              sdram_dqm/output [3:0]  sdram_dqm/' build/ysyxSoCFull.v
 	sed -i 's/wire \[1:0\]  _asic_sdram_dqm/wire [3:0]  _asic_sdram_dqm/' build/ysyxSoCFull.v
+	sed -i 's/ysyx_00000000/ysyx_26010011/' build/ysyxSoCFull.v
 	echo "aaa"
 
 verilog_fix: $(V_FILE_FINAL)
-- 
2.43.0

