From 982928c89f963ef92dfcf51854dc421d1f4da319 Mon Sep 17 00:00:00 2001
From: FZ_XiaoXi <FZXiaoxi@outlook.com>
Date: Wed, 12 Aug 2026 22:41:05 +0800
Subject: [PATCH 09/19] B2 FINISHED!`

---
 rocket-chip | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/rocket-chip b/rocket-chip
index d0c6b50f..2b7015ab 160000
--- a/rocket-chip
+++ b/rocket-chip
@@ -1 +1 @@
-Subproject commit d0c6b50fdefcdbe121e9788433ea51f7efaf1d32
+Subproject commit 2b7015ab88382a524f33033b4a3b2d0f8a9422c9
-- 
2.43.0

