From 6f37882aeb3c337dd2cbd9eca364a874f6193151 Mon Sep 17 00:00:00 2001
From: FZ_XiaoXi <FZXiaoxi@outlook.com>
Date: Mon, 17 Aug 2026 09:32:42 +0800
Subject: [PATCH 16/19] fix bug

---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 4f3c05e9..a52e2943 100644
--- a/Makefile
+++ b/Makefile
@@ -27,6 +27,7 @@ verilog: verilog_fix
 	sed -i 's/output \[1:0\]  sdram_ba/output [2:0]  sdram_ba/' build/ysyxSoCFull.v
 	sed -i 's/wire \[1:0\]  _asic_sdram_ba/wire [2:0]  _asic_sdram_ba/' build/ysyxSoCFull.v
 	sed -i 's/output \[1:0\]  sdram_bundle_ba/output [2:0]  sdram_bundle_ba/' build/ysyxSoCFull.v
+	sed -i 's|Memory\[0:2047\];|Memory\[0:2047\] /*verilator public*/;|' build/ysyxSoCFull.v
 	sed -i 's/ysyx_00000000/ysyx_26010011/' build/ysyxSoCFull.v
 	echo "aaa"
 
-- 
2.43.0

